#Орищенко Виталий
#@Vitosh992

In [2]:
import pandas as pd
import numpy as np

In [5]:
df = pd.read_csv("Titanic-Dataset.csv")
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [16]:
print("Количество пустых ячеек в датасете: " + str(df.isnull().sum().sum()))

Количество пустых ячеек в датасете: 866


In [17]:
print(df.isnull().sum()) # Определим в каких столбцах есть пропуски

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


Так как пустые значения находятся только в ячейках возраста, номера кают и название порта посадки, то для каждого случая выполним замену отдельно.
Для возраста можно использовать замену средним по столбцу, а номера кают и порта посадки будем заменять по предыдущему значению в выборке. А для случаев, если предыдущего значения нет - будем определять по следующему.

In [35]:
df['Age'] = df['Age'].fillna(df['Age'].mean())
df['Cabin'] = df['Cabin'].fillna(method = 'ffill')
df['Cabin'] = df['Cabin'].fillna(method = 'bfill')
df['Embarked'] = df['Embarked'].fillna(method = 'ffill')
df['Embarked'] = df['Embarked'].fillna(method = 'bfill')

In [36]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ...        ...   
886                              Montvila, Rev. Juozas    male  27.000000   
887                       Graham, Miss. Margaret Edith  female  19.000000   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.699118   
889                              Behr, Mr. Karl Howell    male  26.000000   
890                                Dooley, Mr. Patrick    male  32.000000   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
0        1      0         A/5 21171   7.2500   C85        S  
1        1      0          PC 17599  71.2833   C85        C  
2        0      0  STON/O2. 3101282   7.9250   C85        S  
3        1      0            113803  53.1000  C123        S  
4        0      0            373450   8.0500  C123        S  
..     ...    ...               ...      ...   ...      ...  
886      0      0            211536  13.0000   C50        S  
887      0      0            112053  30.0000   B42        S  
888      1      2        W./C. 6607  23.4500   B42        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500  C148        Q  

[891 rows x 12 columns]

In [37]:
# Проверяем заменили ли мы все ячейки
print("Количество пустых ячеек в датасете: " + str(df.isnull().sum().sum()))

Количество пустых ячеек в датасете: 0


Перед нахождением максимального, минимального значения, количества уникальных значений для каждого признака проведем первичный анализ данных. PassengerId - это идентификатор именно для выборки, поэтому смысла анализировать его нет (мин = 1, макс = 891, уникальных значений = 891).

Параметр Survived показывает выжил ли пассажир. Принимает булево-значение в целочисленном представлении (1 или 0), соотственно, находить его параметры тоже нет смысла.

Для параметры Pclass имеет смысл найти параметры, однако, они будут говорить только о количестве классов у пассажиров.

In [50]:
print("Минимальное значение столбца Pclass: " + str(df['Pclass'].min()))
print("Максимальное значение столбца Pclass: " + str(df['Pclass'].max()))
print("Количество уникальных значений Pclass: " + str(df['Pclass'].nunique()))

Минимальное значение столбца Pclass: 1
Максимальное значение столбца Pclass: 3
Количество уникальных значений Pclass: 3


Параметр Name представляет собой информацию об имени пассажиров. Так как тип данных строка, следовательно, мы не можем найти минимум и максимум. Найдем только количество уникальных значений (но оно, конечно, равно всей длину датафрейма)

In [51]:
print("Количество уникальных значений Name: " + str(df['Name'].nunique()))

Количество уникальных значений Name: 891


Поле Sex аналогично полю Survived.

Для поля Age мы можем найти все необходимые статистики.

In [52]:
print("Минимальное значение столбца Age: " + str(df['Age'].min()))
print("Максимальное значение столбца Age: " + str(df['Age'].max()))
print("Количество уникальных значений Age: " + str(df['Age'].nunique()))

Минимальное значение столбца Age: 0.42
Максимальное значение столбца Age: 80.0
Количество уникальных значений Age: 89


Такая же ситуация и с параметром SibSp и Parch

In [53]:
print("Минимальное значение столбца SibSp: " + str(df['SibSp'].min()))
print("Максимальное значение столбца SibSp: " + str(df['SibSp'].max()))
print("Количество уникальных значений SibSp: " + str(df['SibSp'].nunique()))

Минимальное значение столбца SibSp: 0
Максимальное значение столбца SibSp: 8
Количество уникальных значений SibSp: 7


In [54]:
print("Минимальное значение столбца Parch: " + str(df['Parch'].min()))
print("Максимальное значение столбца Parch: " + str(df['Parch'].max()))
print("Количество уникальных значений Parch: " + str(df['Parch'].nunique()))

Минимальное значение столбца Parch: 0
Максимальное значение столбца Parch: 6
Количество уникальных значений Parch: 7


С параметром Ticket также, как и с параметром Name.

In [55]:
print("Количество уникальных значений Ticket: " + str(df['Ticket'].nunique()))

Количество уникальных значений Ticket: 681


Для параметра Fare найдем все статистики

In [56]:
print("Минимальное значение столбца Fare: " + str(df['Fare'].min()))
print("Максимальное значение столбца Fare: " + str(df['Fare'].max()))
print("Количество уникальных значений Fare: " + str(df['Fare'].nunique()))

Минимальное значение столбца Fare: 0.0
Максимальное значение столбца Fare: 512.3292
Количество уникальных значений Fare: 248


Для Cabin и Embarked, также как и с Name и Ticket, найдем количество уникальных значений.

In [57]:
print("Количество уникальных значений Cabin: " + str(df['Cabin'].nunique()))

Количество уникальных значений Cabin: 147


In [58]:
print("Количество уникальных значений Embarked: " + str(df['Embarked'].nunique()))

Количество уникальных значений Embarked: 3


Для каждого численного параметры (Pclass, Age, SibSp, Parch, Fare) найдем коэффициент корреляции Пирсона в паре с параметров Survived. Определим таким образом наиболее значимые параметры и направленность тенденции.

In [63]:
print(np.corrcoef(df['Survived'],df['Pclass'])[0,1]) #Pclass
print(np.corrcoef(df['Survived'],df['Age'])[0,1]) #Age
print(np.corrcoef(df['Survived'],df['SibSp'])[0,1]) #SibSp
print(np.corrcoef(df['Survived'],df['Parch'])[0,1]) #Parch
print(np.corrcoef(df['Survived'],df['Fare'])[0,1]) #Fare

-0.3384810359610148
-0.06980851528714314
-0.03532249888573564
0.08162940708348371
0.2573065223849625


На основании полученных результатов можно сделать вывод, что чем ниже был класс пассажира, тем более выше была вероятность выжить. Аналогичная ситуация с возрастом, но влияние не такое сильное. Число братьев, сестер или супруг влияло наименьшим образом на вероятность выжить. Количество детей или родителей повышало шансы на спасение, как и стоимость билета, однако последний параметр влиял гораздо сильнее.